src/protocols/secure_channel/CASESession.cpp

```diff
@@ -339,57 +300,54 @@ CHIP_ERROR CASESession::DeriveSecureSession(SecureSession & session, SecureSessi
+                                              kP256_PublicKey_Length + sizeof(uint64_t) * 4);
 
     System::PacketBufferTLVWriter tlvWriter;
     System::PacketBufferHandle msg_R1;
     TLV::TLVType outerContainerType                      = TLV::kTLVType_NotSpecified;
     uint8_t initiatorRandom[kSigmaParamRandomNumberSize] = { 0 };
+    uint8_t destinationIdentifier[kSHA256_Hash_Length]   = { 0 };
 
-    msg_R1 = System::PacketBufferHandle::New(data_len);
-    VerifyOrReturnError(!msg_R1.IsNull(), CHIP_ERROR_NO_MEMORY);
-
-    // Step 1
-    // Fill in the random value
-    ReturnErrorOnFailure(DRBG_get_bytes(initiatorRandom, kSigmaParamRandomNumberSize));
-
-// Step 4
+    // Generate an ephemeral keypair
 #ifdef ENABLE_HSM_CASE_EPHERMAL_KEY
     mEphemeralKey.SetKeyId(CASE_EPHEMERAL_KEY);
 #endif
     ReturnErrorOnFailure(mEphemeralKey.Initialize());
 
-    // Start writing TLV
+    // Fill in the random value
+    ReturnErrorOnFailure(DRBG_get_bytes(initiatorRandom, kSigmaParamRandomNumberSize));
+
+    // Construct Sigma1 Msg
+    msg_R1 = System::PacketBufferHandle::New(data_len);
+    VerifyOrReturnError(!msg_R1.IsNull(), CHIP_ERROR_NO_MEMORY);
+
     tlvWriter.Init(std::move(msg_R1));
     ReturnErrorOnFailure(tlvWriter.StartContainer(TLV::AnonymousTag, TLV::kTLVType_Structure, outerContainerType));
-    ReturnErrorOnFailure(tlvWriter.PutBytes(CASETLVTag::kRandom, initiatorRandom, sizeof(initiatorRandom)));
-
-    // Step 5
-    uint16_t n_trusted_roots = mOpCredSet->GetCertCount();
-    // Initiator's session ID
-    ReturnErrorOnFailure(tlvWriter.Put(CASETLVTag::kSessionID, mConnectionState.GetLocalKeyID(), true));
-    // Step 2/3
-    ReturnErrorOnFailure(tlvWriter.Put(CASETLVTag::kNumberofTrustedRootIDs, n_trusted_roots, true));
-    for (uint16_t i = 0; i < n_trusted_roots; ++i)
+    ReturnErrorOnFailure(tlvWriter.PutBytes(TLV::ContextTag(1), initiatorRandom, sizeof(initiatorRandom)));
+    // Retrieve Session Identifier
+    ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(2), mConnectionState.GetLocalKeyID(), true));
+    // Generate a Destination Identifier
     {
-        CertificateKeyId trustedRootId = mOpCredSet->GetTrustedRootId(i);
-        if (!trustedRootId.empty())
-        {
-            ReturnErrorOnFailure(tlvWriter.PutBytes(CASETLVTag::kTrustedRootID, trustedRootId.data(), trustedRootId.size()));
-        }
+        ChipCertificateData nodeOperationalCertificate;
+        FabricId fabricId;
+        MutableByteSpan destinationIdSpan(destinationIdentifier, sizeof(destinationIdentifier));
+
+        ReturnErrorOnFailure(DecodeChipCert(mOpCredSet->GetDevOpCred(mTrustedRootId), mOpCredSet->GetDevOpCredLen(mTrustedRootId),
+                                            nodeOperationalCertificate));
+        ReturnErrorOnFailure(nodeOperationalCertificate.mSubjectDN.GetCertFabricId(fabricId));
+        ReturnErrorOnFailure(GenerateDestinationID(ByteSpan(initiatorRandom, sizeof(initiatorRandom)), &mTrustedRootId,
```

Comment:
GenerateDestinationID should take-in the IPK and public key of the root cert, rather than directly access the certs. Makes it more testable and less stateful.